### PR TITLE
fix deinflection bug

### PR DESCRIPTION
**Pull request by StefanVukovic99**
With #503, more than one inflection rule chain is now shown. Seems the deinflector is producing an incorrect chain for -te form of ichidan verbs: 
image

Not sure if the two subcategories of ichidan verbs in this PR are aptly named, but this seems to fix the problem.

### CODE REVIEW

**Comment by djahandarie** [comment]
I honestly don't quite understand what's going on in the deinflector logic to determine if this is right (and whether there's a better naming option) so hopefully toasted-nutbread can provide some insight.

**Comment by Scrub1492** [comment]
lgtm

File: test/deinflector.test.js [hunk 1]
```diff
@@ -895,7 +895,9 @@ function testDeinflections() {
                 {term: 'くる', source: 'くさせられない', rule: 'vk'},
 
                 {term: 'かわいい', source: 'かわいげ',   rule: 'adj-i', reasons: ['-ge']},
-                {term: '可愛い',   source: 'かわいげ',   rule: 'adj-i', reasons: ['-ge']}
+                {term: '可愛い',   source: 'かわいげ',   rule: 'adj-i', reasons: ['-ge']},
+
+                {term: '食べる', source: '食べて', reasons: ['-te', 'progressive or perfect', 'masu stem']}
```
Comment (toasted-nutbread) [inline test/deinflector.test.js:900]: Shouldn't rule be `'v1'` here?

**Comment by toasted-nutbread** [comment]
I believe the reason this works is because of some shenanigans with the bitflags and the way that the `Translator` class uses `Deinflector.rulesMatch`. Specifically, this code:

```js
const definitionRules = Deinflector.rulesToRuleFlags(databaseEntry.rules);
for (const deinflection of uniqueDeinflectionArrays[databaseEntry.index]) {
    const deinflectionRules = deinflection.rules;
    if (!partsOfSpeechFilter || Deinflector.rulesMatch(deinflectionRules, definitionRules)) {
        deinflection.databaseEntries.push(databaseEntry);
    }
}
```

`Deinflector.rulesMatch` is only checking to see if _any_ of the bits match between the two rules passed into it, so `v1d` and `v1p` will both be matched by terms with the `v1` flag.

This is seems to be why these rules are set up this way:
```js
        ['v1',    /** @type {import('translation-internal').DeinflectionRuleFlags} */ (0b000000011)], // Verb ichidan
        ['v1d',   /** @type {import('translation-internal').DeinflectionRuleFlags} */ (0b000000010)], // Verb ichidan dictionary form
        ['v1p',   /** @type {import('translation-internal').DeinflectionRuleFlags} */ (0b000000001)], // Verb ichidan progressive or perfect
```

This functions as a way to just prevent this from happening:

```
食べて
  => ('て' => 'てる' via 'masu stem'; rules='v1')
食べてる
  => ('てる' => 'て' via 'progressive or perfect'; rules='iru')
食べて
  => ('て' => 'る' via '-te'; rules='v1')
食べる
```

With the new rules:

```
食べて
  => ('て' => 'てる' via 'masu stem'; rules='v1d')
食べてる
  => attempts: ('てる' => 'て' via 'progressive or perfect')
     However, rulesIn='v1p', which doesn't match 'v1d', so this doesn't work
<stop>
```

**Comment by StefanVukovic99** [comment]
Exactly.

**Comment by toasted-nutbread** [comment]
I think the overall idea is that `v1d` is used as a sort of termination indicator to prevent anything that has been converted from `masu stem` to have a `る` ending, that can no longer be used for the `progressive or perfect` conversion, since that is effectively undoing the addition of the `る`.

Generally I don't see a huge issue with it being done this way, I can't think of any immediate edge cases where this breaks down. The only thing is that it makes the subtleties of the logic a bit more confusing to understand I suppose, due to the idiosyncrasies of the bitflags test I mentioned above, but that's fine. I think that issue can be addressed by adjusting the structure of the deinflect.json file, which I think would also benefit other language support as well. I'll try looking into that at some point.

**Review by toasted-nutbread — APPROVED** [review]

**Review by djahandarie — APPROVED** [review]
Thanks for breaking it down. I think I understand, and approve of the change. That said, this introduces even more complexity into the already confusing logic here, so I definitely feel a need for some rework of the deinflector overall, as well as deinflection rule names to be a little less opaque, with more clarity of intent of each part in the naming and structure.
